qiskit_machine_learning/kernels/quantum_kernel.py

```diff
@@ -0,0 +1,322 @@
+
+        q = QuantumRegister(self._feature_map.num_qubits, 'q')
+        c = ClassicalRegister(self._feature_map.num_qubits, 'c')
+        qc = QuantumCircuit(q, c)
+
+        x_dict = dict(zip(self._feature_map.ordered_parameters, x))
+        psi_x = self._feature_map.assign_parameters(x_dict)
+        qc.append(psi_x.to_instruction(), qc.qubits)
+
+        if not is_statevector_sim:
+            if y is not None and len(y) != self._feature_map.num_parameters:
+                raise ValueError("y and class feature map incompatible dimensions.\n" +
+                                 "y has %s dimensions, but feature map has %s." %
+                                 (len(y), self._feature_map.num_parameters))
+
+            if y is None:
+                y = x
+
+            y_dict = dict(zip(self._feature_map.ordered_parameters, y))
+            psi_y_dag = self._feature_map.assign_parameters(y_dict)
+            qc.append(psi_y_dag.to_instruction().inverse(), qc.qubits)
+
+            if measurement:
+                qc.barrier(q)
+                qc.measure(q, c)
+        return qc
+
+    def _compute_overlap(self, idx, results, is_statevector_sim, measurement_basis):
+        """
+        Helper function to compute overlap for given input.
+        """
+        if is_statevector_sim:
+            # |<0|Psi^dagger(y) x Psi(x)|0>|^2, take the amplitude
+            v_a, v_b = [results.get_statevector(int(i)) for i in idx]
+            tmp = np.vdot(v_a, v_b)
+            kernel_value = np.vdot(tmp, tmp).real  # pylint: disable=no-member
+        else:
+            result = results.get_counts(idx)
+
+            kernel_value = result.get(measurement_basis, 0) / sum(result.values())
+        return kernel_value
+
+    def evaluate(self, x_vec: np.ndarray, y_vec: np.ndarray = None) -> np.ndarray:
+        r"""
+        Construct kernel matrix for given data and feature map
+
+        If y is None, self inner product is calculated.
+        If using `statevector_simulator`, only build circuits for :math:`\Psi(x)|0\rangle`,
+        then perform inner product classically.
+
+        Args:
+            x_vec: 1D or 2D array of datapoints, NxD, where N is the number of datapoints,
+                                                            D is the feature dimension
+            y_vec: 1D or 2D array of datapoints, MxD, where M is the number of datapoints,
+                                                            D is the feature dimension
+
+        Returns:
+            2D matrix, NxM
+
+        Raises:
+            AlgorithmError:
+                - A quantum instance or backend has not been provided to the class
+            ValueError:
+                - x_vec and/or y_vec are not one or two dimensional arrays
+                - x_vec and/or y_vec have incompatible dimension with feature map
+        """
+        if self._quantum_instance is None:
+            raise AlgorithmError("A QuantumInstance or Backend "
```

Comment:
There is a QiskitMachineLearningError - it may be preferable to use that instead of the Algorithm one

Reply:
Changed from `AlgorithmError` to `QiskitMachineLearningError`